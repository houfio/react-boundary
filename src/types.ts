import { ErrorInfo } from 'react';

export type State = {
  error?: Error,
  errorInfo?: ErrorInfo
};

export type Hook = [Error | undefined, ErrorInfo | undefined];
