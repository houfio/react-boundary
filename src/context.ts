import { createContext } from 'react';

import { State } from './types';

export const context = createContext<State>({});
