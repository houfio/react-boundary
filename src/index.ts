export { boundary } from './boundary';
export { useError } from './useError';
