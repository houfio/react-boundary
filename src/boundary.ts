import { Component, ComponentType, createElement, ErrorInfo } from 'react';

import { context } from './context';
import { State } from './types';

export function boundary<P extends object>(component: ComponentType<P>) {
  return class Boundary extends Component<P, State> {
    public static displayName = `Boundary(${component.displayName || component.name || 'Component'})`;

    public state = {};

    public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
      this.setState({
        error,
        errorInfo
      });
    }

    public render() {
      return createElement(context.Provider, {
        value: this.state,
        children: createElement(component, this.props)
      });
    }
  };
}
