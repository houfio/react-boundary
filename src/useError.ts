import { useContext } from 'react';

import { context } from './context';
import { Hook } from './types';

export function useError(): Hook {
  const { error, errorInfo } = useContext(context);

  return [error, errorInfo];
}
