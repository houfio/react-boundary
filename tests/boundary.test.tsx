import * as React from 'react';
import { render } from 'react-testing-library';

import { boundary, useError } from '../src';

type Props = {
  crash: boolean
};

const Component = boundary(({ crash }: Props) => {
  const [error, info] = useError();

  if (crash && !error) {
    throw new Error('CRASHED!');
  }

  return error && info ? (
    <span>
      Oh no! Something happened: {error.message}
      {info.componentStack}
    </span>
  ) : (
    <span>
      Everything is fine
    </span>
  );
});

it('renders the component', () => {
  const { asFragment } = render(
    <Component crash={false}/>
  );

  expect(asFragment()).toMatchSnapshot();
});

it('renders the errored component', () => {
  const { asFragment } = render(
    <Component crash={true}/>
  );

  expect(asFragment()).toMatchSnapshot();
});
