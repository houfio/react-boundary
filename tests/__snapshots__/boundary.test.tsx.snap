// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<DocumentFragment>
  <span>
    Everything is fine
  </span>
</DocumentFragment>
`;

exports[`renders the errored component 1`] = `
<DocumentFragment>
  <span>
    Oh no! Something happened: CRASHED!
    in Unknown (created by Boundary(Component))
    in Boundary(Component)
  </span>
</DocumentFragment>
`;
